external/mnist_data.cmake

```diff
@@ -0,0 +1,50 @@
+# -----------------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2016, Numenta, Inc.  Unless you have purchased from
+# Numenta, Inc. a separate commercial license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# -----------------------------------------------------------------------------
+
+# Fetch MNIST dataset from online archive
+#
+if(EXISTS ${REPOSITORY_DIR}/build/ThirdParty/share/eigen-master.zip)
+    set(URL ${REPOSITORY_DIR}/build/ThirdParty/share/eigen-master.zip)
+else()
+    set(URL "https://github.com/wichtounet/mnist/archive/master.zip")
+    set(HASH "0bcecef3aaed7c619f0baac02697a76dff3da54f0c17f1aaaa92a2928ca300f3")
+endif()
+
+message(STATUS "obtaining MNIST data")
+include(DownloadProject/DownloadProject.cmake)
+download_project(PROJ mnist
+	            PREFIX ${EP_BASE}/mnist_data
+		    URL    ${URL}
+		    URL_HASH SHA256=${HASH}
+                    UPDATE_DISCONNECTED 1
+		    #           QUIET
+   )	
+# No build. This is a data only package
+add_subdirectory(${mnist_SOURCE_DIR}/example/ ${mnist_BINARY_DIR})
+set(mnist_INCLUDE_DIRS ${mnist_SOURCE_DIR}/include/mnist)
+if (MSVC)
+set(mnist_LIBRARIES   "${mnist_BINARY_DIR}$<$<CONFIG:Release>:/Release/mnist.lib>$<$<CONFIG:Debug>:/Debug/mnist.lib>")
```

Comment:
If this is a data only download then there should not be a .lib So these lines can be removed.

Reply:
The add_subdirectory( ) call does build an example executable.  I am assuming that you do not want that. So you don't need the add_subdirectory line.

In fact, I think the only thing you need after the download_project is the export of the mnist_SOURCE_DIR path.
```
FILE(APPEND "${EXPORT_FILE_NAME}" "mnist_SOURCE_DIR@@@${mnist_SOURCE_DIR}\n")
```

Reply:
> In fact, I think the only thing you need after the download_project is the export of the mnist_SOURCE_DIR path. 

which approach do you find better? 
- just download the data files? I'd choose that if I could download directly those: http://yann.lecun.com/exdb/mnist/ 

- but DownloadProject had some issues with the gz (?), so I opted for the git repo download (contains data + parsing library) -> in which case we might like to build the lib and use it, instead of our custom methods for parsing the binary data. 

What do you think?

Reply:
> issues with gz?  

That is its most often used mode. What issues did you have?
All you should need is
- The check to see if it is manually downloaded vs a url
- The DownloadProject to download and unwrap it.
- A line of code to export the path so it can be found by other cmake's outside of external.

Now if you need other parsers that adds more lines.

Reply:
> That is its most often used mode 

so it does support gz? 

> manually downloaded vs a url 

I ignored manual, focused only on URL, which showed progress downloading

> and unwrap it. 

failed to extract, as it named the .gz file as .tar, somehow? 

I'll try to replicate that here on CI, so you can see

Reply:
I can take a look at this.  A .gz file is a compressed .tar so it must do this with a two step process...decompress the .gz and then expand the .tar.  CMake should take care of all of that.